### PR TITLE
Allow for different colored fill

### LINKED ISSUE #37: Allow different fill color to spark line color
As title suggests, it should be possible to get the fill drawing with a different color to the main line.

I have managed to get this working locally, with a subclass of spark view, and wondered if it would be worth porting it back into Spark and submitting a PR?

### PR DESCRIPTION
**Pull request by RichyHBM**
Fixes #37 
Adds a new attribute allowing you to set a different color for the fill, it also fixes the animators to work with fill

### CODE REVIEW

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 1]
```diff
@@ -164,13 +169,22 @@ private void init(Context context, AttributeSet attrs, int defStyleAttr, int def
...
 
+        sparkLinePaint.setStyle(Paint.Style.STROKE);
         sparkLinePaint.setColor(lineColor);
         sparkLinePaint.setStrokeWidth(lineWidth);
         sparkLinePaint.setStrokeCap(Paint.Cap.ROUND);
         if (cornerRadius != 0) {
             sparkLinePaint.setPathEffect(new CornerPathEffect(cornerRadius));
         }
 
+        sparkFillPaint.setColor(fillColor);
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: Might be simpler to do:
```
sparkFillPaint.set(sparkLinePaint);
```

and then change only what's needed.

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 2]
```diff
@@ -224,6 +238,7 @@ private void populatePath() {
 
         // make our main graph path
         sparkPath.reset();
+        sparkFillPath.reset();
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: I'm not sure about the animations yet, but I'm pretty sure that for just regular drawing we only need one path. We can draw the path twice, once with each paint.

In fact, I think if we have these separate, and somebody has set data boundaries to 'zoom out' a bit from the sparkline, you'll be able to see that the stroke only shows on the top of the fill shape. But I feel like we'd probably want the stroke to go around the entire path in this case

File: spark/src/main/java/com/robinhood/spark/animation/LineSparkAnimator.java [hunk 3]
```diff
@@ -39,10 +41,31 @@ public void onAnimationUpdate(ValueAnimator animation) {
...
+
                 pathMeasure.getSegment(0, animatedPathLength, linePath, true);
+                fillPath.addPath(linePath);
+
+                float lastPos[] = {0f, 0f};
+                //get point at animatedValue
+                pathMeasure.getPosTan(pathMeasure.getLength() * animatedValue, lastPos, null);
+
+                // if we're filling the graph in, close the path's circuit
+                final Float fillEdge = sparkView.getFillEdge();
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/animation/LineSparkAnimator.java]: I'd probably argue we should not try to animate the fill for this animation, since the path is unclosed until the very end of it. If anything, I'd say we wait until the stroke is done animating and then fade in the fill color.

**Comment by RichyHBM** [comment]
I think the main problem with merging them is that then the spark line would draw an outline to the fill, if you can think of a way to do it do let me know though and I'll change it!

**Comment by danh32** [comment]
I'm not entirely sure what you mean. Are you saying for folks who want to have a fill color without a stroke color? I think the path should be the same regardless of how we choose to draw it.

**Comment by RichyHBM** [comment]
Yes, so if its the same path when we draw the line it will always have an outline when also drawing with a fill, with multiple paths you can have the line drawing one color and the fill another without having an outline around all the fill

**Comment by danh32** [comment]
So the stroke vs fill is a property of the paint, not the path.

So I'm thinking in all cases we should have a single path representing the sparkline. Then we can do something like:

```
canvas.drawPath(path, strokePaint)
if (shouldFill) {
    canvas.drawPath(path, fillPaint)
}
```

**Comment by RichyHBM** [comment]
Yes, but to draw the fill, the path needs to be closed which would mean the stroke paint would always draw an outline (drawing the spark line, and then a line down to 0, and back to the start of the spark)

**Comment by danh32** [comment]
Yeah, I think that's what we'd want in this case. I think it would look really odd to just have the stroke on the top of the graph instead of all around it. So for fill cases, the sparkpath will be closed, and both filled and stroked.

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 4]
```diff
@@ -267,7 +275,7 @@ private void populatePath() {
         invalidate();
     }
 
-    private Float getFillEdge() {
+    public Float getFillEdge() {
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: i don't think i like this being public. since it's only being used in the animation, i think we can revert the changes to the animations and revert this change as well.

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 5]
```diff
@@ -426,6 +455,14 @@ public void setSparkAnimator(SparkAnimator sparkAnimator) {
         this.sparkAnimator = sparkAnimator;
     }
 
+    /**
+     * ScaleHelper class
+     * @return a {@link ScaleHelper} or null
+     */
+    public ScaleHelper getScaleHelper() {
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: i definitely don't like this becoming public. same deal as above, we can revert if we revert the animation changes.

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 6]
```diff
@@ -707,7 +762,7 @@ private void clearData() {
     /**
      * Helper class for handling scaling logic.
      */
-    static class ScaleHelper {
+    public static class ScaleHelper {
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: revert

File: spark/src/main/java/com/robinhood/spark/animation/LineSparkAnimator.java [hunk 7]
```diff
@@ -19,6 +19,7 @@ public Animator getAnimation(final SparkView sparkView) {
 
         final Path linePath = sparkView.getSparkLinePath();
 
+
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/animation/LineSparkAnimator.java]: revert

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 8]
```diff
@@ -354,6 +367,11 @@ public void setPadding(int left, int top, int right, int bottom) {
     protected void onDraw(Canvas canvas) {
         super.onDraw(canvas);
         canvas.drawPath(baseLinePath, baseLinePaint);
+
+        if(fillType != FillType.NONE && (sparkAnimator == null || sparkAnimator.hasFinishedAnimating())){
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: the draw loop shouldn't know about the animation at all, beyond just drawing the renderPath. let's take those conditions out.

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 9]
```diff
@@ -387,6 +421,7 @@ public float getLineWidth() {
     public void setLineWidth(float lineWidth) {
         this.lineWidth = lineWidth;
         sparkLinePaint.setStrokeWidth(lineWidth);
+        sparkFillPaint.setStrokeWidth(lineWidth);
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: i don't think sparkFillPaint needs a stroke width, it'll always be hidden by the stroke drawn by the sparkLinePaint. we should probably reset the stroke width to zero in the constructor as well, since we're copying values over when setting up the sparkFillPaint

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 10]
```diff
@@ -426,6 +463,14 @@ public void setSparkAnimator(SparkAnimator sparkAnimator) {
         this.sparkAnimator = sparkAnimator;
     }
 
+    /**
+     * ScaleHelper class
+     * @return a {@link ScaleHelper} or null
+     */
+    ScaleHelper getScaleHelper() {
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: I really don't want to expose the ScaleHelper at all.

File: spark/src/main/java/com/robinhood/spark/animation/MorphSparkAnimator.java [hunk 11]
```diff
@@ -74,4 +85,5 @@ public void setDuration(long duration) {
         animator.setDuration(duration);
     }
 
+    public Boolean hasFinishedAnimating() { return ended; }
```
Comment (naturalwarren) [inline spark/src/main/java/com/robinhood/spark/animation/MorphSparkAnimator.java]: Small question, why did we move from `boolean` to `Boolean` here?

**Comment by RichyHBM** [comment]
That would mean the animators would be completely broken with fills, just to confirm, you are OK with this?

File: spark/src/main/java/com/robinhood/spark/animation/MorphSparkAnimator.java (hunk 11 above)
Comment (danh32) [reply spark/src/main/java/com/robinhood/spark/animation/MorphSparkAnimator.java]: good catch! i think we want to revert this method entirely.

**Comment by danh32** [comment]
They're completely broken currently as well, right? Given that it's not a regression, I'm pretty comfortable with it. I'm open to a separate diff that fixes them in some way, but I'm thinking that's out of scope for this PR.

**Comment by RichyHBM** [comment]
OK, will revert my changes today!

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 12]
```diff
@@ -202,6 +212,13 @@ protected void onSizeChanged(int w, int h, int oldW, int oldH) {
         populatePath();
     }
 
+    /**
+     * Populates the {@linkplain #sparkPath} with points
+     */
+    public void rebuildPath() {
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: i don't think we should be exposing this

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 13]
```diff
@@ -331,9 +348,6 @@ public Path getSparkLinePath() {
     public void setAnimationPath(final Path animationPath) {
         this.renderPath.reset();
         this.renderPath.addPath(animationPath);
-        this.renderPath.rLineTo(0, 0);
-
-        invalidate();
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: i'm pretty concerned about these deletions.

File: spark/src/main/java/com/robinhood/spark/animation/LineSparkAnimator.java [hunk 14]
```diff
@@ -11,14 +12,13 @@
  * Animates the sparkline by path-tracing from the first point to the last.
  */
 public class LineSparkAnimator implements SparkAnimator {
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/animation/LineSparkAnimator.java]: let's completely revert this file

File: spark/src/main/java/com/robinhood/spark/animation/MorphSparkAnimator.java [hunk 15]
```diff
@@ -57,13 +57,18 @@ public void onAnimationUpdate(ValueAnimator animation) {
 
                 // set the updated path for the animation
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/animation/MorphSparkAnimator.java]: and this file

File: spark/src/main/java/com/robinhood/spark/animation/SparkAnimator.java [hunk 16]
```diff
@@ -18,5 +18,4 @@
      * @param sparkView The SparkView object
      */
     Animator getAnimation(final SparkView sparkView);
-
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/animation/SparkAnimator.java]: and this one

File: spark/src/main/java/com/robinhood/spark/SparkView.java (hunk 12 above)
Comment (RichyHBM) [reply spark/src/main/java/com/robinhood/spark/SparkView.java]: Agree, will change this

File: spark/src/main/java/com/robinhood/spark/SparkView.java (hunk 13 above)
Comment (RichyHBM) [reply spark/src/main/java/com/robinhood/spark/SparkView.java]: I'm a little unsure why the LineTo 0,0 is there, but happy to change it back

File: spark/src/main/java/com/robinhood/spark/SparkView.java [hunk 17]
```diff
@@ -490,12 +523,12 @@ public void setFillType(@FillType int fillType) {
 
             switch (fillType) {
                 case FillType.NONE:
-                    sparkLinePaint.setStyle(Paint.Style.STROKE);
+                    sparkFillPaint.setStyle(Paint.Style.STROKE);
```
Comment (danh32) [inline spark/src/main/java/com/robinhood/spark/SparkView.java]: I don't think this is necessary anymore. This used to toggle the singular sparkLinePaint between fill and non-fill modes. Since fill and non-fill have their own paints now, I think this whole switch can go away.
